.teamcity/src/main/kotlin/configurations/Gradleception.kt

```diff
@@ -20,25 +26,67 @@ class Gradleception(model: CIBuildModel, stage: Stage) : BaseGradleBuildType(sta
         javaCrash = false
     }
 
+    params {
+        // Override the default commit id so the build steps produce reproducible distribution
+        param("env.BUILD_COMMIT_ID", "HEAD")
+    }
+
+    /*
+     To avoid unnecessary rerun, what we do here is a bit complicated:
+
+     1. Build a Gradle distribution with a fixed timestamp and hash it, but never use this distribution.
+     2. Build a Gradle distribution with this hash as a version + fixed timestamp -> dogfood-first
+     3. Build a Gradle distribution using dogfood-first with the "hash of dogfood-first" as a version + fixed timestamp different from the one above -> dogfood-second
+     4. Use dogfood-second to run a Gradle build which runs tests.
+
+     */
+    val dogfoodTimestamp1 = "20210101010101+0000"
+    val dogfoodTimestamp2 = "20210202020202+0000"
     val buildScanTagForType = buildScanTag("Gradleception")
     val defaultParameters = (buildToolGradleParameters() + listOf(buildScanTagForType) + "-Porg.gradle.java.installations.auto-download=false").joinToString(separator = " ")
 
-    applyDefaults(model, this, ":distributions-full:install", notQuick = true, extraParameters = "-Pgradle_installPath=dogfood-first $buildScanTagForType", extraSteps = {
-        localGradle {
-            name = "BUILD_WITH_BUILT_GRADLE"
-            tasks = "clean :distributions-full:install"
-            gradleHome = "%teamcity.build.checkoutDir%/dogfood-first"
-            gradleParams = "-Pgradle_installPath=dogfood-second -PignoreIncomingBuildReceipt=true $defaultParameters"
-        }
-        localGradle {
-            name = "QUICKCHECK_WITH_GRADLE_BUILT_BY_GRADLE"
-            tasks = "clean sanityCheck test"
-            gradleHome = "%teamcity.build.checkoutDir%/dogfood-second"
-            gradleParams = defaultParameters
-        }
-    })
+    applyDefaults(
+        model,
+        this,
+        ":distributions-full:install",
+        notQuick = true,
+        extraParameters = "-Pgradle_installPath=dogfood-first-for-hash -PignoreIncomingBuildReceipt=true -PbuildTimestamp=$dogfoodTimestamp1 $buildScanTagForType",
+        extraSteps = {
+            calculateMd5AndSetEnvStep("CALCULATE_MD5_VERSION_FOR_FIRST_DOGFOODING_DISTRIBUTION", "%teamcity.build.checkoutDir%/dogfood-first-for-hash")
+            gradleWrapper {
+                name = "BUILD_GRADLE_DISTRIBUTION"
+                tasks = "clean :distributions-full:install"
+                gradleParams = "-Pgradle_installPath=dogfood-first -PignoreIncomingBuildReceipt=true -PbuildTimestamp=$dogfoodTimestamp1 $defaultParameters"
+            }
+
+            calculateMd5AndSetEnvStep("CALCULATE_MD5_VERSION_FOR_SECOND_DOGFOODING_DISTRIBUTION", "%teamcity.build.checkoutDir%/dogfood-first")
```

Comment:
Why do we need to do this again here? We only need to calculate the MD5 sum once on the distribution we never use, and use two different timestamps for dogfood1 & dogfood2, right?

Reply:
Right. Fixed.